server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java

```diff
@@ -822,32 +822,39 @@ public List<Round> getDownsamplingRoundsFor(
      * the backing indices that are managed by the data stream lifecycle".
      */
     public List<Index> getNonWriteIndicesOlderThan(
-        TimeValue age,
+        TimeValue retentionPeriod,
         Function<String, IndexMetadata> indexMetadataSupplier,
         @Nullable Predicate<IndexMetadata> indicesPredicate,
         LongSupplier nowSupplier
     ) {
         List<Index> olderIndices = new ArrayList<>();
         for (Index index : indices) {
-            IndexMetadata indexMetadata = indexMetadataSupplier.apply(index.getName());
-            if (indexMetadata == null) {
-                // we would normally throw exception in a situation like this however, this is meant to be a helper method
-                // so let's ignore deleted indices
-                continue;
-            }
-            TimeValue indexLifecycleDate = getGenerationLifecycleDate(indexMetadata);
-            if (indexLifecycleDate != null) {
-                long nowMillis = nowSupplier.getAsLong();
-                if (nowMillis >= indexLifecycleDate.getMillis() + age.getMillis()) {
-                    if (indicesPredicate == null || indicesPredicate.test(indexMetadata)) {
-                        olderIndices.add(index);
-                    }
-                }
+            if (isIndexOderThan(index, retentionPeriod.getMillis(), nowSupplier.getAsLong(), indicesPredicate, indexMetadataSupplier)) {
+                olderIndices.add(index);
             }
         }
         return olderIndices;
```

Comment:
Nit: I usually think the `Stream` API is a bit more elegant. I don't feel strong enough about it to actually "request changes", but thought I'd mention it. AKA, up to you :).

Reply:
Thanks Niels, I agree, I was just trying to balance performance and elegance. Let me explain, I think in the stream approach we create:
- Here we create one list in which we add elements and we use two iterators.
- In the stream approach, we would first need to concatenate the two lists into a new list, or convert them into streams and then concatenate these streams. Both felt a bit too much overhead.

But I do not feel too strongly about it. Happy to revert back to streams :). What do you think?

Reply:
Ahh I hadn't realized that you're going to add the failure store indices afterwards as well of course. Maybe, when you actually implement that part, I'd play around with having a local `Consumer<Index>` variable that you can then use to `indices.forEach(consumer)` and `failureStoreIndices.forEach(consumer)`. But not sure if that ends up being (more) elegant or not. I definitely agree with the setup you have now so I'll resolve this comment, thanks :)

Reply:
Interesting idea, let's work on it on the follow up :)